### ISSUE TITLE
Pynids limitation with stream reassembly

### ISSUE THREAD

**Issue by jared703** [MITRECND/chopshop#62]
Team,

Is there any plans to move away from Pynids in the future? The lack of error handling with Pynids leaves it difficult in packetloss scenerio(s) as the stream is ignored completely if Pynids can't reassemble it.

Best,
JG

**Comment by Mraoul** [comment]
I actually really dislike nids/pynids -- it was the requirement for chopshop when it was first created since pynids was heavily used in my shop and others as the basis for C2 protocol decoders, but it's always come with limitations and bugs. The issue in finding another back-end, though, is that it has to have significant improvements over nids to warrant the switch over since it could possibly break backwards compatibility (something I have been trying to maintain as much as possible). I've been following the progress of libuinet as a possible replacement but that project doesn't seem to be at a  a state where I'd feel comfortable using it to replace nids/pynids.

If you happen to know of any projects that could potentially replace nids, I'd be very interested in taking a look at them.

**Comment by jared703** [comment]
Thanks for the quick reply. I'll certainly be on the lookout.

**Comment by comrumino** [comment]
Hi @Mraoul
I started work on adding ChopShop to ArchStrike. Since Python 2 is hitting end of life in 2020, Pynids must be ported to Python 3. Before putting more time and effort into my [fork](https://github.com/comrumino/pynids), it would be nice to have to have an update on the current road map for ChopShop. Of course, I could also put my time and effort in elsewhere if that would be more helpful.

As of now, my fork does compile and a functional [PKGBUILD](https://github.com/ArchStrike/ArchStrike/blob/master/pending/python-pynids/PKGBUILD ). I do need to review my porting changes; I kind of just added a mess of macro's until it stopped throwing errors---my main concern is really bytes vs string differences.

Lmk if I should create an issue with Pynids too.
**tl;dr**

1. If I put the effort into porting Pynids to Python 3, does it still have a future as a dependency of ChopShop? Side note, a Python 3 Pynids would be nice so a PR could be made to a branch other than master until the port has been thoroughly vetted.
2. My [fork](https://github.com/comrumino/pynids) of Pynids compiles and requires improvement.
3. Should I create a related issue under [Pynids](https://github.com/MITRECND/pynids) too?
